src/main/cpp/lang/xml.cpp

```diff
@@ -98,6 +141,434 @@ struct XML_Highlighter : Highlighter_Base {
     }
 
 private:
+    bool expect_content_spec()
+    {
+        advance(match_whitespace(remainder));
+
+        if (remainder.starts_with(element_content_spec_empty)) {
+            emit_and_advance(element_content_spec_empty.size(), Highlight_Type::keyword);
+            return true;
+        }
+
+        if (remainder.starts_with(element_content_spec_any)) {
+            emit_and_advance(element_content_spec_any.size(), Highlight_Type::keyword);
+            return true;
+        }
+
+        if (remainder.starts_with(u8"(#PCDATA")) {
+            emit_and_advance(1, Highlight_Type::symbol_punc);
+            emit_and_advance(7, Highlight_Type::keyword);
+        }
+
+        static constexpr std::array<char8_t, 8> non_name_chars
+            = { u8'(', u8')', u8'|', u8'*', u8'+', u8'?', u8'>', ',' };
+
+        constexpr auto is_after_name = [](std::u8string_view str) {
+            return std::ranges::find(non_name_chars, str.front()) != std::end(non_name_chars)
+                || match_whitespace(str);
+        };
+
+        advance(match_whitespace(remainder));
+        while (!remainder.starts_with(u8'>') && !remainder.empty()) {
+            auto current = remainder[0];
+            if (std::ranges::find(non_name_chars, current) != std::end(non_name_chars)) {
+                emit_and_advance(1, Highlight_Type::symbol_punc);
+            }
+            else if (expect_name(Highlight_Type::name, is_after_name)) {
+            }
+            else if (std::size_t white_space_len = match_whitespace(remainder)) {
+                advance(white_space_len);
+            }
+            else {
+                return true;
+            }
+        }
+
+        return true;
+    }
+
+    bool expect_default_att_decl()
+    {
+        if (remainder.starts_with(u8"#REQUIRED")) {
+            emit_and_advance(9, Highlight_Type::keyword);
+            return true;
+        }
+
+        if (remainder.starts_with(u8"#IMPLIED")) {
+            emit_and_advance(8, Highlight_Type::keyword);
+            return true;
+        }
+
+        if (remainder.starts_with(u8"#FIXED")) {
+            emit_and_advance(6, Highlight_Type::keyword);
+            advance(match_whitespace(remainder));
+        }
+
+        return expect_attribute_value();
```

Comment:
This is problematic because it means you could already emit the tokens for `#FIXED` but then return `false`, indicating that nothing was matched.

`expect_*` functions should be "all or nothing", i.e. only return `false` if nothing was emitted.

A possible solution is to just say that once you've matched `#FIXED`, the result is `true` no matter whether you actually find an attribute value or not.